src/typechecker.erl

```diff
@@ -3410,6 +3546,46 @@ refine_ty(Ty1, Ty2, TEnv) ->
         _NotDisjoint -> throw(no_refinement) %% imprecision
     end.
 
+-spec refine_map_field_ty({_, _}) -> gradualizer_type:abstract_type().
+%% For the same key K in both M1 and M2 the diff over its field is:
+%% M1 \ M2 |  :=  |  =>
+%% --------+------+------
+%%    :=   | none |  =>
+%%    =>   |  #{} | none
```

Comment:
Nice table, but I don't understand exactly what you mean. Is the below equivalent?

```
%% For the same key K in both M1 and M2 the diff over its field (ignoring other keys) is:
%%
%% M1 \ M2    | #{K := V2} | #{K => V2}
%% -----------+------------+---------------------
%% #{K := V1} | none       | #{K => diff(V1, V2)}
%% #{K => V1} | #{}        | none
```

Reply:
Yes, with the difference that currently the implementation of the top right is:

```erlang
refine_map_field_ty({?type(map_field_exact, KVTy), ?type(map_field_assoc, KVTy)}) ->
    %% M1 = #{x := y}
    %% M2 = #{x => y}
    %% M1 \ M2 = #{x => y}
    [type(map_field_assoc, KVTy)];
 ```
 
So it's just a change of the field "exactness", not of the value type. Should it affect the value type? I don't think so, since `y \ y = none()`, which would give a `x => none()` field, and that doesn't make much sense for variant elimination from a type that's checked for exhaustiveness.

Reply:
So to fix the table:

```
%% For the same key K in both M1 and M2 the diff over its field (ignoring other keys) is:
%%
%% M1 \ M2    | #{K := V} | #{K => V}
%% -----------+-----------+--------------
%% #{K := V}  | none      | #{K => V}
%% #{K => V}  | #{}       | none
```

Reply:
In yet other words, the maps M1 and M2 that we subtract are different because of the set of all fields in each one, but the fields that overlap are required to match with regard to both key and value types.

Reply:
Ok, so you'll update these comments in any way you like? Then I guess we can merge this, unless there's anything else left to do, is there?

Reply:
Updated :+1: I think this one is ready to merge.